### PR TITLE
Feat/improv decoding

### PR DESCRIPTION
**Pull request by clemlak**
Improves the encoding / decoding by using Assembly

### CODE REVIEW

**Comment by clemlak** [comment]
- Add `SWAP_ASSET` and `SWAP_QUOTE` opcodes
- Merge (encode / decode)-(allocate / deallocate) to save some space, since they were the exact same thing

File: contracts/libraries/AssemblyLib.sol [hunk 1]
```diff
@@ -312,4 +312,32 @@ library AssemblyLib {
...
+                        value := shl(8, value)
+                        length := sub(length, 1)
+                    }
+            }
+
+            mstore(output, length)
+            mstore(add(0x20, output), value)
+        }
+    }
+    */
```
Comment (Alexangelj) [inline contracts/libraries/AssemblyLib.sol:342]: Should delete or fix soon!

File: contracts/libraries/FVMLib.sol [hunk 2]
```diff
@@ -185,42 +185,76 @@ function encodeJumpInstruction(bytes[] memory instructions)
...
 
-function decodePairIdFromPoolId(uint64 poolId) pure returns (uint24) {
-    return uint24(poolId >> 40);
+/**
+ * @dev Decodes the `pair id` from a `pool id`.
+ * @param poolId Pool id to use for the decoding
+ * @return pairId Corresponding pair id
+ * @custom:example
+ * ```
+ * uint24 pairId = decodePairIdFromPoolId(46183783333895);
```
Comment (Alexangelj) [inline contracts/libraries/FVMLib.sol:194]: example missing

File: contracts/libraries/FVMLib.sol [hunk 3]
```diff
@@ -185,42 +185,76 @@ function encodeJumpInstruction(bytes[] memory instructions)
...
+ * `0x | pairId (24 bits) | isMutable (8 bits) | poolNonce (32 bits)`
+ * @param data Encoded pool id
+ * @return poolId Pool id converted from bytes to uint64
+ * @return pairId Pair id of the pool
+ * @return isMutable True if the pool is mutable
+ * @return poolNonce Pool nonce of the pool
+ * @custom:example
+ * ```
+ * (uint64 poolId, uint24 pairId, uint8 isMutable, uint32 poolNonce)
+ *     = decodePoolId(0x000007010000002a);
```
Comment (Alexangelj) [inline contracts/libraries/FVMLib.sol:238]: missing example

File: contracts/libraries/FVMLib.sol [hunk 4]
```diff
@@ -328,68 +378,60 @@ function decodeCreatePool(bytes calldata data)
...
-    priorityFee = uint16(bytes2(data[24:26]));
-    fee = uint16(bytes2(data[26:28]));
-    vol = uint16(bytes2(data[28:30]));
-    dur = uint16(bytes2(data[30:32]));
-    jit = uint16(bytes2(data[32:34]));
-    uint8 pointer0 = uint8(bytes1(data[34]));
-    maxPrice = AssemblyLib.toAmount(data[35:pointer0]);
-    price = AssemblyLib.toAmount(data[pointer0:]);
+    assembly {
+        pairId := shr(232, calldataload(add(1, data.offset)))
```
Comment (Alexangelj) [inline contracts/libraries/FVMLib.sol:382]: Might be good to store the bits i.e. `232` in constants

**Review by Alexangelj — APPROVED** [review]
this is awesome! LGTM

**Pull request merged**